src/IECoreScene/MeshAlgoTangents.cpp

```diff
@@ -239,3 +261,221 @@ std::pair<PrimitiveVariable, PrimitiveVariable> IECoreScene::MeshAlgo::calculate
 
 	return std::make_pair( tangentPrimVar, bitangentPrimVar );
 }
+
+std::pair<PrimitiveVariable, PrimitiveVariable> IECoreScene::MeshAlgo::calculateTangentsFromPrimitiveCentroid(
+	const MeshPrimitive *mesh,
+	const std::string &position /* = "P" */,
+	const std::string &normal /* = "N" */,
+	bool orthoTangents,
+	bool leftHanded
+)
+{
+	// get point and normal data
+	const V3fVectorData *positionData = mesh->variableData<V3fVectorData>( position );
+	const V3fVectorData *normalData = mesh->variableData<V3fVectorData>( normal );
+	if( !positionData )
+	{
+		std::string e = boost::str( boost::format( "MeshAlgo::calculateTangentsFromPrimitiveCentroid : MeshPrimitive has no Vertex \"%s\" primitive variable." ) % position );
+		throw InvalidArgumentException( e );
+	}
+	if( !normalData )
+	{
+		std::string e = boost::str( boost::format( "MeshAlgo::calculateTangentsFromPrimitiveCentroid : MeshPrimitive has no Vertex \"%s\" primitive variable." ) % normal );
+		throw InvalidArgumentException( e );
+	}
+	const V3fVectorData::ValueType &points = positionData->readable();
+	const V3fVectorData::ValueType &normals = normalData->readable();
```

Comment:
need to double check that the normals are `Interpolation::Vertex` as well, since you're assuming the same size as the positions